# **Подготовка и выявление проблем исходного набора данных**

Перед реализацией алгоритмов машинного обучения была проводится процедура инжиниринга данных.

In [ ]:
import pandas as pd
import gdown
url = 'https://drive.google.com/uc?id=1nytZOYptCgRVEfxH1qX53sS_XAG9MFv6'  # Замените FILE_ID на идентификатор вашего файла
output = 'dataset.csv'
gdown.download(url, output, quiet=False)

Downloading...
From: https://drive.google.com/uc?id=1nytZOYptCgRVEfxH1qX53sS_XAG9MFv6
To: /content/dataset.csv
100%|██████████| 5.96k/5.96k [00:00<00:00, 11.8MB/s]


'dataset.csv'

In [ ]:
ds = pd.read_csv(output)
ds.head()

№\№   pd1   pd2   pd3  opt1  opt2  opt3  opt4  opt5  opt6  ...  res16  \
0  Т1  1.10  7.52  0.97  1280  1480  1360  0.74  0.71  0.71  ...      0   
1  Т2  0.57  3.74  0.91  2050   707  1230  0.99  0.89  0.94  ...      0   
2  Т3  1.10  7.52  0.16  1350  1150  1210  1.71  1.70  1.73  ...      0   
3  Т4  0.80  5.33  1.13   368   530   390  1.78  1.72  1.74  ...      0   
4  Т5  0.80  5.33  1.69   101   125    14  1.55  1.41  1.46  ...      0   

   res17  res18  res19  res20  res21  res22  res23  res24  res25  
0      0      0      0      1      1      1      1      1      1  
1      0      0      0      1      1      0      1      1      1  
2      0      0      0      1      1      0      1      1      1  
3      0      0      0      0      1      0      1      1      1  
4      0      0      0      0      1      0      1      1      1  

[5 rows x 56 columns]

In [ ]:
ds = ds.iloc[1:, 1:]
ds.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 25 entries, 1 to 25
Data columns (total 55 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   pd1     25 non-null     float64
 1   pd2     25 non-null     float64
 2   pd3     25 non-null     float64
 3   opt1    25 non-null     int64  
 4   opt2    25 non-null     int64  
 5   opt3    25 non-null     int64  
 6   opt4    25 non-null     float64
 7   opt5    25 non-null     float64
 8   opt6    25 non-null     float64
 9   opt7    25 non-null     float64
 10  opt8    25 non-null     float64
 11  opt9    25 non-null     float64
 12  opt10   25 non-null     float64
 13  opt11   25 non-null     float64
 14  opt12   25 non-null     float64
 15  opt13   25 non-null     float64
 16  opt14   25 non-null     float64
 17  opt15   25 non-null     float64
 18  opt16   25 non-null     float64
 19  opt17   25 non-null     float64
 20  opt18   25 non-null     float64
 21  opt19   25 non-null     float64
 22  opt2

In [ ]:
ds.describe()

pd1        pd2        pd3         opt1         opt2         opt3  \
count  25.000000  25.000000  25.000000    25.000000    25.000000    25.000000   
mean    0.716400   4.774000   1.040000  1611.960000  1248.160000  1301.560000   
std     0.200018   1.420337   0.511468  2028.496703  1409.214488  1176.375014   
min     0.300000   1.870000   0.160000    29.000000    83.000000    14.000000   
25%     0.570000   3.740000   0.600000   291.000000   323.000000   390.000000   
50%     0.790000   5.320000   1.030000   735.000000   853.000000  1210.000000   
75%     0.800000   5.330000   1.500000  1912.000000  1465.000000  1558.000000   
max     1.100000   7.520000   1.930000  8390.000000  6500.000000  4600.000000   

            opt4       opt5       opt6       opt7  ...  res16  res17  res18  \
count  25.000000  25.000000  25.000000  25.000000  ...   25.0   25.0   25.0   
mean    1.432400   1.022000   1.025600   1.010000  ...    0.0    0.0    0.0   
std     1.612716   0.283387   0.292006   0.979847  ...    0.0    0.0    0.0   
min     0.550000   0.470000   0.500000   0.020000  ...    0.0    0.0    0.0   
25%     0.900000   0.880000   0.840000   0.360000  ...    0.0    0.0    0.0   
50%     1.150000   0.980000   0.970000   0.610000  ...    0.0    0.0    0.0   
75%     1.320000   1.140000   1.110000   1.210000  ...    0.0    0.0    0.0   
max     9.040000   1.720000   1.740000   3.680000  ...    0.0    0.0    0.0   

       res19      res20  res21  res22      res23  res24  res25  
count   25.0  25.000000   25.0   25.0  25.000000  25.00  25.00  
mean     0.0   0.440000    1.0    0.0   0.640000   0.96   0.96  
std      0.0   0.506623    0.0    0.0   0.489898   0.20   0.20  
min      0.0   0.000000    1.0    0.0   0.000000   0.00   0.00  
25%      0.0   0.000000    1.0    0.0   0.000000   1.00   1.00  
50%      0.0   0.000000    1.0    0.0   1.000000   1.00   1.00  
75%      0.0   1.000000    1.0    0.0   1.000000   1.00   1.00  
max      0.0   1.000000    1.0    0.0   1.000000   1.00   1.00  

[8 rows x 55 columns]

Проанализировав исходный набор данных, был сделан вывод, о необходимости проведения серии исследовательских экспериментов, связанных с подготовкой исходного набора данных, обучением и оценкой моделей.
Общая гипотеза экспериментов заключается в том, что от выбора обучающих и целевых признаков будет изменяться качество обучения моделей машинного обучения. При этом важно учитывать, что под качеством подразумевается не только достижение наилучших значений метрик оценки качества обучения, но и то, как конкретно влияют выбранные обучающие признаки на целевую для данной работы задачу классификации состояния изоляции силового масляного трансформатора.


**Подготовка к эксперименту №1**

Цель эксперимента — проверить, могут ли модели машинного обучения эффективно прогнозировать состояние изоляции силового трансформатора, используя ограниченный набор данных (паспортные данные, визуальные осмотры, электротехнические измерения для 27 трансформаторов) без фильтрации признаков по их значимости. Эксперимент поможет подтвердить возможность применения ИИ в диагностике трансформаторов с минимальными исходными данными, определит подходящие алгоритмы и повысит надежность диагностики в условиях ограниченности данных. Ожидается, что некоторые модели смогут эффективно работать даже при минимальной предобработке, а результаты дадут рекомендации для дальнейшего применения ИИ в диагностике электрооборудования.

В процессе подготовки решались проблемы низкого качества данных (пропуски и выбросы), дисбаланса классов и несоответствия масштабов признаков. Были применены методы предобработки: заполнение пропусков, удаление выбросов, настройка гиперпараметров для балансировки классов и масштабирование данных с помощью библиотек Python. Выводы указывают на необходимость увеличения данных, балансировки классов и использования подходящих метрик для повышения точности модели.

**Подготовка к эксперименту №2**

Цель эксперимента — выявить обучающие признаки из ограниченного набора данных, которые дадут наилучшие результаты в прогнозировании электрических характеристик силового масляного трансформатора. Ожидается, что удаление части признаков повлияет на качество обучения моделей, предоставив основы для оценки методов мониторинга и диагностики. Эксперимент покажет практическую применимость выбранных признаков и их влияние на избежание мультиколлинеарности.

Для предотвращения шума и мультиколлинеарности были удалены такие признаки, как «Год изготовления», «Установленная мощность, кВА», «Пиковая мощность, кВА», «ПБН. Повреждение контактов», «ПБН. Отсутствие неповрежденных ответвлений», «Вводы. Повреждение уплотнений», «Радиаторы. Механическое повреждение», «Внешний осмотр воздухоосушителя. Отсутствие предусмотренного проектом воздухоосушителя», «Воздухоосушитель. Отсутствие», «Радиаторы. Отсутствие», и все признаки, связанные с внешним осмотром оборудования, так как они описывают внешние характеристики, не влияющие напрямую на прогнозирование электрических параметров.

# **Экспериментальная проверка базовых алгоритмов на ограниченном наборе входных данны**

Процесс разработки модели начинается с подготовки исходного набора данных, учитывая проблемы и особенности, описанные в подготовке для экспериментов №1 и №2. Далее данные загружаются в формате CSV, подготавливаются и делятся на обучающую и тестовую выборки (тренировочная — 20%). Затем данные масштабируются с помощью функций StandardScaler() и scaler.transform() из библиотеки sklearn.

После преобразования данных модели обучаются с предварительной настройкой гиперпараметров. Корректируется дисбаланс классов и проводится их оптимизация с помощью алгоритма GridSearchCV, исследующего все комбинации гиперпараметров. Для оценки качества обучения применяются метрики: точность (accuracy), точность (precision), полнота (recall), F-мера (F1-score) и сводные метрики sklearn — Micro Average, Macro Average, Weighted Average и Samples Average. Weighted avg лучше подходит для задач с несбалансированными классами, учитывая их дисбаланс, а macro avg применяется, если классы сбалансированы, а samples avg используется для задач с многозначными метками.

**Процедурпа разработки моделей МО с применением различных алгоритмов**

**Алгоритм**: Логистическая регрессия


In [ ]:
import pandas as pd
from sklearn.ensemble import RandomForestClassifier
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score, mean_squared_error, classification_report
from sklearn.preprocessing import StandardScaler

X = ds.drop(ds.loc[:, 'res1':'res25'], axis=1)
y = ds.loc[:, 'res1':'res25']

# Разобъем наш DF на тестовую и обучающую выборку, используем ф-цию train_test_split()
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.20, random_state=1)
X_train.shape, X_test.shape, y_train.shape, y_test.shape

#Масштабирование данных
numeric = ['pd1', 'pd2', 'pd3', 'opt1', 'opt2', 'opt3',
       'opt4', 'opt5', 'opt6', 'opt7', 'opt8', 'opt9', 'opt10', 'opt11',
       'opt12', 'opt13', 'opt14', 'opt15', 'opt16', 'opt17', 'opt18', 'opt19',
       'opt20', 'opt21', 'opt22', 'opt23', 'opt24', 'opt25', 'opt26', 'opt27']
scaler = StandardScaler()
scaler.fit(X_test[numeric])
X_train[numeric] = scaler.transform(X_train[numeric])
X_test[numeric] = scaler.transform(X_test[numeric])

random_forest = RandomForestClassifier(n_estimators=100, max_depth=7, class_weight='balanced')
random_forest.fit(X_train, y_train)
y_pred_rf = random_forest.predict(X_test)
random_forest.score(X_train, y_train)

random_forest_train = round(random_forest.score(X_train, y_train) * 100, 2)
random_forest_accuracy = round(accuracy_score(y_pred_rf, y_test) * 100, 2)

print("Training Accuracy    :",random_forest_train ,"%")
print("Model Accuracy Score :",random_forest_accuracy ,"%")
print("\033[1m--------------------------------------------------------\033[0m")
print("Classification_Report: \n",classification_report(y_test,y_pred_rf))
print("\033[1m--------------------------------------------------------\033[0m")


Training Accuracy    : 100.0 %
Model Accuracy Score : 20.0 %
--------------------------------------------------------
Classification_Report: 
               precision    recall  f1-score   support

           0       1.00      1.00      1.00         5
           1       0.50      1.00      0.67         2
           2       0.00      0.00      0.00         0
           3       0.00      0.00      0.00         0
           4       1.00      1.00      1.00         5
           5       1.00      1.00      1.00         5
           6       0.00      0.00      0.00         0
           7       0.00      0.00      0.00         0
           8       0.00      0.00      0.00         0
           9       0.60      1.00      0.75         3
          10       0.00      0.00      0.00         0
          11       0.00      0.00      0.00         1
          12       0.00      0.00      0.00         0
          13       1.00      1.00      1.00         5
          14       0.00      0.00      0.00   

/usr/local/lib/python3.10/dist-packages/sklearn/metrics/_classification.py:1531: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", len(result))
/usr/local/lib/python3.10/dist-packages/sklearn/metrics/_classification.py:1531: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 in labels with no true samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", len(result))
/usr/local/lib/python3.10/dist-packages/sklearn/metrics/_classification.py:1531: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 in labels with no true nor predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", len(result))


Вариация с применением grid search (в результате Best Parameters : {'max_depth': None, 'max_features': 'auto', 'min_samples_leaf': 4, 'min_samples_split': 10, 'n_estimators': 300})

In [ ]:
import pandas as pd
from sklearn.ensemble import RandomForestClassifier
from sklearn.model_selection import train_test_split, GridSearchCV
from sklearn.metrics import accuracy_score, classification_report
from sklearn.preprocessing import StandardScaler

# Разделение на признаки и целевые переменные
X = ds.drop(ds.loc[:, 'res1':'res25'], axis=1)
y = ds.loc[:, 'res1':'res25']

# Разделение на обучающую и тестовую выборки
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.20, random_state=1)

# Масштабирование данных
numeric = ['pd1', 'pd2', 'pd3', 'opt1', 'opt2', 'opt3',
           'opt4', 'opt5', 'opt6', 'opt7', 'opt8', 'opt9', 'opt10', 'opt11',
           'opt12', 'opt13', 'opt14', 'opt15', 'opt16', 'opt17', 'opt18', 'opt19',
           'opt20', 'opt21', 'opt22', 'opt23', 'opt24', 'opt25', 'opt26', 'opt27']
scaler = StandardScaler()
scaler.fit(X_train[numeric])
X_train[numeric] = scaler.transform(X_train[numeric])
X_test[numeric] = scaler.transform(X_test[numeric])

# Создание базовой модели RandomForestClassifier
random_forest = RandomForestClassifier(class_weight='balanced')

# Задание сетки гиперпараметров для поиска
param_grid = {
    'n_estimators': [100, 200, 300],
    'max_depth': [None, 10, 20, 30],
    'min_samples_split': [2, 5, 10],
    'min_samples_leaf': [1, 2, 4],
    'max_features': ['auto', 'sqrt', 'log2']
}

# Применение GridSearchCV
grid_search = GridSearchCV(estimator=random_forest, param_grid=param_grid, cv=3, n_jobs=-1, verbose=2)
grid_search.fit(X_train, y_train)

# Лучшая модель и параметры
best_params = grid_search.best_params_
best_model = grid_search.best_estimator_

# Обучение лучшей модели на обучающих данных
best_model.fit(X_train, y_train)
y_pred_rf = best_model.predict(X_test)

# Оценка модели
random_forest_train = round(best_model.score(X_train, y_train) * 100, 2)
random_forest_accuracy = round(accuracy_score(y_pred_rf, y_test) * 100, 2)

print("Best Parameters      :", best_params)
print("Training Accuracy    :", random_forest_train, "%")
print("Model Accuracy Score :", random_forest_accuracy, "%")
print("\033[1m--------------------------------------------------------\033[0m")
print("Classification Report: \n", classification_report(y_test, y_pred_rf))
print("\033[1m--------------------------------------------------------\033[0m")


Fitting 3 folds for each of 324 candidates, totalling 972 fits


/usr/local/lib/python3.10/dist-packages/sklearn/model_selection/_validation.py:540: FitFailedWarning: 
324 fits failed out of a total of 972.
The score on these train-test partitions for these parameters will be set to nan.
If these failures are not expected, you can try to debug them by setting error_score='raise'.

Below are more details about the failures:
--------------------------------------------------------------------------------
324 fits failed with the following error:
Traceback (most recent call last):
  File "/usr/local/lib/python3.10/dist-packages/sklearn/model_selection/_validation.py", line 888, in _fit_and_score
    estimator.fit(X_train, y_train, **fit_params)
  File "/usr/local/lib/python3.10/dist-packages/sklearn/base.py", line 1466, in wrapper
    estimator._validate_params()
  File "/usr/local/lib/python3.10/dist-packages/sklearn/base.py", line 666, in _validate_params
    validate_parameter_constraints(
  File "/usr/local/lib/python3.10/dist-packages/sklearn/util

Best Parameters      : {'max_depth': None, 'max_features': 'sqrt', 'min_samples_leaf': 2, 'min_samples_split': 10, 'n_estimators': 100}
Training Accuracy    : 25.0 %
Model Accuracy Score : 0.0 %
--------------------------------------------------------
Classification Report: 
               precision    recall  f1-score   support

           0       1.00      1.00      1.00         5
           1       0.00      0.00      0.00         2
           2       0.00      0.00      0.00         0
           3       0.00      0.00      0.00         0
           4       1.00      1.00      1.00         5
           5       1.00      1.00      1.00         5
           6       0.00      0.00      0.00         0
           7       0.00      0.00      0.00         0
           8       0.00      0.00      0.00         0
           9       0.60      1.00      0.75         3
          10       0.00      0.00      0.00         0
          11       0.20      1.00      0.33         1
          12       0

/usr/local/lib/python3.10/dist-packages/sklearn/metrics/_classification.py:1531: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", len(result))
/usr/local/lib/python3.10/dist-packages/sklearn/metrics/_classification.py:1531: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 in labels with no true samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", len(result))
/usr/local/lib/python3.10/dist-packages/sklearn/metrics/_classification.py:1531: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 in labels with no true nor predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", len(result))


**Алгоритм**: Метод XGBoost

In [ ]:
import xgboost as xgb
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score, mean_squared_error, classification_report
from sklearn.preprocessing import StandardScaler

X = ds.drop(ds.loc[:, 'res1':'res25'], axis=1)
y = ds.loc[:, 'res1':'res25']

# Разобъем наш DF на тестовую и обучающую выборку, используем ф-цию train_test_split()
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.20, random_state=1)
X_train.shape, X_test.shape, y_train.shape, y_test.shape

#Масштабирование данных
numeric = ['pd1', 'pd2', 'pd3', 'opt1', 'opt2', 'opt3',
       'opt4', 'opt5', 'opt6', 'opt7', 'opt8', 'opt9', 'opt10', 'opt11',
       'opt12', 'opt13', 'opt14', 'opt15', 'opt16', 'opt17', 'opt18', 'opt19',
       'opt20', 'opt21', 'opt22', 'opt23', 'opt24', 'opt25', 'opt26', 'opt27']
scaler = StandardScaler()
scaler.fit(X_test[numeric])
X_train[numeric] = scaler.transform(X_train[numeric])
X_test[numeric] = scaler.transform(X_test[numeric])


xgb = xgb.XGBClassifier(random_state=12345, class_weight='balanced')
xgb.fit(X_train,y_train)

y_pred_xgb = xgb.predict(X_test)
xgb.score(X_train, y_train)


xgb_train = round(xgb.score(X_train, y_train) * 100, 2)
xgb_accuracy = round(accuracy_score(y_pred_xgb, y_test) * 100, 2)

print("Training Accuracy    :",xgb_train ,"%")
print("Model Accuracy Score :",xgb_accuracy ,"%")
print("\033[1m--------------------------------------------------------\033[0m")
print("Classification_Report: \n",classification_report(y_test,y_pred_xgb))
print("\033[1m--------------------------------------------------------\033[0m")

/usr/local/lib/python3.10/dist-packages/xgboost/core.py:158: UserWarning: [09:50:39] WARNING: /workspace/src/learner.cc:740: 
Parameters: { "class_weight" } are not used.

  warnings.warn(smsg, UserWarning)


Training Accuracy    : 85.0 %
Model Accuracy Score : 0.0 %
--------------------------------------------------------
Classification_Report: 
               precision    recall  f1-score   support

           0       1.00      1.00      1.00         5
           1       0.50      1.00      0.67         2
           2       0.00      0.00      0.00         0
           3       0.00      0.00      0.00         0
           4       1.00      1.00      1.00         5
           5       1.00      1.00      1.00         5
           6       0.00      0.00      0.00         0
           7       0.00      0.00      0.00         0
           8       0.00      0.00      0.00         0
           9       0.75      1.00      0.86         3
          10       0.00      0.00      0.00         0
          11       0.00      0.00      0.00         1
          12       0.00      0.00      0.00         0
          13       1.00      1.00      1.00         5
          14       0.00      0.00      0.00     

/usr/local/lib/python3.10/dist-packages/sklearn/metrics/_classification.py:1531: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", len(result))
/usr/local/lib/python3.10/dist-packages/sklearn/metrics/_classification.py:1531: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 in labels with no true samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", len(result))
/usr/local/lib/python3.10/dist-packages/sklearn/metrics/_classification.py:1531: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 in labels with no true nor predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", len(result))


вариант с применением grid search

In [ ]:
import xgboost as xgb
import pandas as pd
from sklearn.model_selection import train_test_split, GridSearchCV
from sklearn.metrics import accuracy_score, classification_report
from sklearn.preprocessing import StandardScaler

# Разделение на признаки и целевые переменные
X = ds.drop(ds.loc[:, 'res1':'res25'], axis=1)
y = ds.loc[:, 'res1':'res25']

# Разделение на обучающую и тестовую выборки
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.20, random_state=1)
X_train.shape, X_test.shape, y_train.shape, y_test.shape

# Масштабирование данных
numeric = ['pd1', 'pd2', 'pd3', 'opt1', 'opt2', 'opt3',
           'opt4', 'opt5', 'opt6', 'opt7', 'opt8', 'opt9', 'opt10', 'opt11',
           'opt12', 'opt13', 'opt14', 'opt15', 'opt16', 'opt17', 'opt18', 'opt19',
           'opt20', 'opt21', 'opt22', 'opt23', 'opt24', 'opt25', 'opt26', 'opt27']
scaler = StandardScaler()
scaler.fit(X_train[numeric])
X_train[numeric] = scaler.transform(X_train[numeric])
X_test[numeric] = scaler.transform(X_test[numeric])

# Создание базовой модели XGBClassifier
xgb_model = xgb.XGBClassifier(random_state=12345)

# Задание сетки гиперпараметров для поиска
param_grid = {
    'n_estimators': [100, 200, 300],
    'max_depth': [3, 5, 7],
    'learning_rate': [0.01, 0.1, 0.2],
    'subsample': [0.8, 0.9, 1.0],
    'colsample_bytree': [0.8, 0.9, 1.0],
    'scale_pos_weight': [1, 2, 3]  # Параметр для баланса классов
}


# Применение GridSearchCV
grid_search = GridSearchCV(estimator=xgb_model, param_grid=param_grid, cv=3, n_jobs=-1, verbose=2)
grid_search.fit(X_train, y_train)

# Лучшая модель и параметры
best_params = grid_search.best_params_
best_model = grid_search.best_estimator_

# Обучение лучшей модели на обучающих данных
best_model.fit(X_train, y_train)
y_pred_xgb = best_model.predict(X_test)

# Оценка модели
xgb_train = round(best_model.score(X_train, y_train) * 100, 2)
xgb_accuracy = round(accuracy_score(y_pred_xgb, y_test) * 100, 2)

print("Best Parameters      :", best_params)
print("Training Accuracy    :", xgb_train, "%")
print("Model Accuracy Score :", xgb_accuracy, "%")
print("\033[1m--------------------------------------------------------\033[0m")
print("Classification Report: \n", classification_report(y_test, y_pred_xgb))
print("\033[1m--------------------------------------------------------\033[0m")


Fitting 3 folds for each of 729 candidates, totalling 2187 fits
Best Parameters      : {'colsample_bytree': 0.8, 'learning_rate': 0.1, 'max_depth': 3, 'n_estimators': 100, 'scale_pos_weight': 2, 'subsample': 0.8}
Training Accuracy    : 85.0 %
Model Accuracy Score : 0.0 %
--------------------------------------------------------
Classification Report: 
               precision    recall  f1-score   support

           0       1.00      1.00      1.00         5
           1       0.50      1.00      0.67         2
           2       0.00      0.00      0.00         0
           3       0.00      0.00      0.00         0
           4       1.00      1.00      1.00         5
           5       1.00      1.00      1.00         5
           6       0.00      0.00      0.00         0
           7       0.00      0.00      0.00         0
           8       0.00      0.00      0.00         0
           9       0.75      1.00      0.86         3
          10       0.00      0.00      0.00        

/usr/local/lib/python3.10/dist-packages/sklearn/metrics/_classification.py:1531: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", len(result))
/usr/local/lib/python3.10/dist-packages/sklearn/metrics/_classification.py:1531: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 in labels with no true samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", len(result))
/usr/local/lib/python3.10/dist-packages/sklearn/metrics/_classification.py:1531: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 in labels with no true nor predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", len(result))


Модель с подобранными идеальными параметрами (Best Parameters : {'colsample_bytree': 0.8, 'learning_rate': 0.1, 'max_depth': 3, 'n_estimators': 100, 'scale_pos_weight': 2, 'subsample': 1.0})

In [ ]:
import xgboost as xgb
import pandas as pd
from sklearn.model_selection import train_test_split, GridSearchCV
from sklearn.metrics import accuracy_score, classification_report
from sklearn.preprocessing import StandardScaler

# Разделение на признаки и целевые переменные
X = ds.drop(ds.loc[:, 'res1':'res25'], axis=1)
y = ds.loc[:, 'res1':'res25']

# Разделение на обучающую и тестовую выборки
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.20, random_state=1)
X_train.shape, X_test.shape, y_train.shape, y_test.shape

# Масштабирование данных
numeric = ['pd1', 'pd2', 'pd3', 'opt1', 'opt2', 'opt3',
           'opt4', 'opt5', 'opt6', 'opt7', 'opt8', 'opt9', 'opt10', 'opt11',
           'opt12', 'opt13', 'opt14', 'opt15', 'opt16', 'opt17', 'opt18', 'opt19',
           'opt20', 'opt21', 'opt22', 'opt23', 'opt24', 'opt25', 'opt26', 'opt27']
scaler = StandardScaler()
scaler.fit(X_train[numeric])
X_train[numeric] = scaler.transform(X_train[numeric])
X_test[numeric] = scaler.transform(X_test[numeric])

# Создание базовой модели XGBClassifier
xgb_model = xgb.XGBClassifier(random_state=12345)

# Задание сетки гиперпараметров для поиска
param_grid = {
    'n_estimators': [100],
    'max_depth': [3],
    'learning_rate': [0.1],
    'subsample': [1.0],
    'colsample_bytree': [0.8],
    'scale_pos_weight': [2]  # Параметр для баланса классов
}

# Применение GridSearchCV
grid_search = GridSearchCV(estimator=xgb_model, param_grid=param_grid, cv=3, n_jobs=-1, verbose=2)
grid_search.fit(X_train, y_train)

# Лучшая модель и параметры
best_params = grid_search.best_params_
best_model = grid_search.best_estimator_

# Обучение лучшей модели на обучающих данных
best_model.fit(X_train, y_train)
y_pred_xgb = best_model.predict(X_test)

# Оценка модели
xgb_train = round(best_model.score(X_train, y_train) * 100, 2)
xgb_accuracy = round(accuracy_score(y_pred_xgb, y_test) * 100, 2)

print("Best Parameters      :", best_params)
print("Training Accuracy    :", xgb_train, "%")
print("Model Accuracy Score :", xgb_accuracy, "%")
print("\033[1m--------------------------------------------------------\033[0m")
print("Classification Report: \n", classification_report(y_test, y_pred_xgb))
print("\033[1m--------------------------------------------------------\033[0m")


Fitting 3 folds for each of 1 candidates, totalling 3 fits
Best Parameters      : {'colsample_bytree': 0.8, 'learning_rate': 0.1, 'max_depth': 3, 'n_estimators': 100, 'scale_pos_weight': 2, 'subsample': 1.0}
Training Accuracy    : 85.0 %
Model Accuracy Score : 0.0 %
--------------------------------------------------------
Classification Report: 
               precision    recall  f1-score   support

           0       1.00      1.00      1.00         5
           1       0.33      0.50      0.40         2
           2       0.00      0.00      0.00         0
           3       0.00      0.00      0.00         0
           4       1.00      1.00      1.00         5
           5       1.00      1.00      1.00         5
           6       0.00      0.00      0.00         0
           7       0.00      0.00      0.00         0
           8       0.00      0.00      0.00         0
           9       0.75      1.00      0.86         3
          10       0.00      0.00      0.00         0
  

/usr/local/lib/python3.10/dist-packages/sklearn/metrics/_classification.py:1531: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", len(result))
/usr/local/lib/python3.10/dist-packages/sklearn/metrics/_classification.py:1531: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 in labels with no true samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", len(result))
/usr/local/lib/python3.10/dist-packages/sklearn/metrics/_classification.py:1531: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 in labels with no true nor predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", len(result))


**Алгоритм:** Метод опорных векторов (Support Vector Machines)

In [ ]:
from sklearn.svm import SVC
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score, mean_squared_error, classification_report
from sklearn.preprocessing import StandardScaler

X = ds.drop(ds.loc[:, 'res1':'res25'], axis=1)
y_r = ds.loc[:, 'res1':'res25']

y['faultType'] = ''
for i in y_r.columns:
  y['faultType'] += y_r[i].astype(str)

y = y.drop(columns=y.columns.difference(['faultType']))

# Разобъем наш DF на тестовую и обучающую выборку, используем ф-цию train_test_split()
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.20, random_state=1)
X_train.shape, X_test.shape, y_train.shape, y_test.shape

#Масштабирование данных
numeric = ['pd1', 'pd2', 'pd3', 'opt1', 'opt2', 'opt3',
       'opt4', 'opt5', 'opt6', 'opt7', 'opt8', 'opt9', 'opt10', 'opt11',
       'opt12', 'opt13', 'opt14', 'opt15', 'opt16', 'opt17', 'opt18', 'opt19',
       'opt20', 'opt21', 'opt22', 'opt23', 'opt24', 'opt25', 'opt26', 'opt27']
scaler = StandardScaler()
scaler.fit(X_test[numeric])
X_train[numeric] = scaler.transform(X_train[numeric])
X_test[numeric] = scaler.transform(X_test[numeric])

print('X_train', X_train.shape)
print('X_test', X_test.shape)
print('y_train', y_train.shape)
print('y_test', y_test.shape)

svc = SVC(random_state=12345, class_weight='balanced')
svc.fit(X_train, y_train)

y_pred_svc = svc.predict(X_test)

svc_train = round(svc.score(X_train, y_train) * 100, 2)
svc_accuracy = round(accuracy_score(y_pred_svc, y_test) * 100, 2)

print("Training Accuracy    :",svc_train ,"%")
print("Model Accuracy Score :",svc_accuracy ,"%")
print("\033[1m--------------------------------------------------------\033[0m")
print("Classification_Report: \n",classification_report(y_test,y_pred_svc))
print("\033[1m--------------------------------------------------------\033[0m")

X_train (20, 30)
X_test (5, 30)
y_train (20, 1)
y_test (5, 1)
Training Accuracy    : 60.0 %
Model Accuracy Score : 0.0 %
--------------------------------------------------------
Classification_Report: 
                            precision    recall  f1-score   support

1000110000000100000110111       0.00      0.00      0.00       1.0
1000110000010100000110011       0.00      0.00      0.00       1.0
1000110001000100000010111       0.00      0.00      0.00       1.0
1000110001010100000010011       0.00      0.00      0.00       0.0
1000110001110100000010011       0.00      0.00      0.00       0.0
1100110001000100000010111       0.00      0.00      0.00       1.0
1100110001000100000110111       0.00      0.00      0.00       1.0

                 accuracy                           0.00       5.0
                macro avg       0.00      0.00      0.00       5.0
             weighted avg       0.00      0.00      0.00       5.0

--------------------------------------------------------


/usr/local/lib/python3.10/dist-packages/sklearn/utils/validation.py:1339: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
/usr/local/lib/python3.10/dist-packages/sklearn/metrics/_classification.py:1531: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", len(result))
/usr/local/lib/python3.10/dist-packages/sklearn/metrics/_classification.py:1531: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 in labels with no true samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", len(result))
/usr/local/lib/python3.10/dist-packages/sklearn/metrics/_classification.py:1531: UndefinedMetricWarning: Precis

Поиск лучших параметров с помощью grid search и обучение модели с их помощью

```
# This is formatted as code
```



In [ ]:
from sklearn.svm import SVC
import pandas as pd
from sklearn.model_selection import train_test_split, GridSearchCV
from sklearn.metrics import accuracy_score, classification_report
from sklearn.preprocessing import StandardScaler

# Разделение на признаки и целевые переменные
X = ds.drop(ds.loc[:, 'res1':'res25'], axis=1)
y_r = ds.loc[:, 'res1':'res25']

# Создание целевой переменной 'faultType'
y = pd.DataFrame()
y['faultType'] = y_r.astype(str).agg(''.join, axis=1)

# Разделение на обучающую и тестовую выборки
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.20, random_state=1)

# Масштабирование данных
numeric = ['pd1', 'pd2', 'pd3', 'opt1', 'opt2', 'opt3',
           'opt4', 'opt5', 'opt6', 'opt7', 'opt8', 'opt9', 'opt10', 'opt11',
           'opt12', 'opt13', 'opt14', 'opt15', 'opt16', 'opt17', 'opt18', 'opt19',
           'opt20', 'opt21', 'opt22', 'opt23', 'opt24', 'opt25', 'opt26', 'opt27']

scaler = StandardScaler()
scaler.fit(X_train[numeric])
X_train[numeric] = scaler.transform(X_train[numeric])
X_test[numeric] = scaler.transform(X_test[numeric])

print('X_train', X_train.shape)
print('X_test', X_test.shape)
print('y_train', y_train.shape)
print('y_test', y_test.shape)

# Создание модели SVC
svc = SVC(random_state=12345, class_weight='balanced')

# Определение сетки гиперпараметров
param_grid = {
    'C': [0.1, 1, 10, 100],
    'gamma': [1, 0.1, 0.01, 0.001],
    'kernel': ['linear', 'rbf']
}

# Применение GridSearchCV
grid_search = GridSearchCV(estimator=svc, param_grid=param_grid, cv=3, n_jobs=-1, verbose=2)
grid_search.fit(X_train, y_train.values.ravel())  # Использование values.ravel() для преобразования в 1D массив

# Лучшая модель и параметры
best_params = grid_search.best_params_
best_model = grid_search.best_estimator_

# Обучение лучшей модели на обучающих данных
best_model.fit(X_train, y_train.values.ravel())
y_pred_svc = best_model.predict(X_test)

# Оценка модели1
svc_train = round(best_model.score(X_train, y_train.values.ravel()) * 100, 2)
svc_accuracy = round(accuracy_score(y_pred_svc, y_test) * 100, 2)

print("Best Parameters      :", best_params)
print("Training Accuracy    :", svc_train, "%")
print("Model Accuracy Score :", svc_accuracy, "%")
print("\033[1m--------------------------------------------------------\033[0m")
print("Classification Report: \n", classification_report(y_test, y_pred_svc))
print("\033[1m--------------------------------------------------------\033[0m")


X_train (20, 30)
X_test (5, 30)
y_train (20, 1)
y_test (5, 1)
Fitting 3 folds for each of 32 candidates, totalling 96 fits


/usr/local/lib/python3.10/dist-packages/sklearn/model_selection/_split.py:776: UserWarning: The least populated class in y has only 1 members, which is less than n_splits=3.
  warnings.warn(


Best Parameters      : {'C': 0.1, 'gamma': 1, 'kernel': 'rbf'}
Training Accuracy    : 70.0 %
Model Accuracy Score : 0.0 %
--------------------------------------------------------
Classification Report: 
                            precision    recall  f1-score   support

1000110000000000000110111       0.00      0.00      0.00       0.0
1000110000000100000110111       0.00      0.00      0.00       1.0
1000110000010100000110011       0.00      0.00      0.00       1.0
1000110001000100000010111       0.00      0.00      0.00       1.0
1000110001000100000110111       0.00      0.00      0.00       0.0
1100110001000100000010111       0.00      0.00      0.00       1.0
1100110001000100000110111       0.00      0.00      0.00       1.0

                 accuracy                           0.00       5.0
                macro avg       0.00      0.00      0.00       5.0
             weighted avg       0.00      0.00      0.00       5.0

--------------------------------------------------------

/usr/local/lib/python3.10/dist-packages/sklearn/metrics/_classification.py:1531: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", len(result))
/usr/local/lib/python3.10/dist-packages/sklearn/metrics/_classification.py:1531: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 in labels with no true samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", len(result))
/usr/local/lib/python3.10/dist-packages/sklearn/metrics/_classification.py:1531: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", len(result))
/usr/local/lib/python3.10/dist-packages/sklearn/metrics/_

**Алгоритм**: Метод k-ближайших соседей (k-Nearest Neighbors)

In [ ]:
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.neighbors import KNeighborsClassifier
from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score
from sklearn.metrics import mean_squared_error
from time import time
import matplotlib.pyplot as plt
from sklearn.preprocessing import StandardScaler

X = ds.drop(ds.loc[:, 'res1':'res25'], axis=1)
y = ds.loc[:, 'res1':'res25']

# Разделение на тренировочную и валидационную выборки
X_train, X_val, y_train, y_val = train_test_split(X, y, test_size=0.2, random_state=42)

#Масштабирование данных
numeric = ['pd1', 'pd2', 'pd3', 'opt1', 'opt2', 'opt3',
       'opt4', 'opt5', 'opt6', 'opt7', 'opt8', 'opt9', 'opt10', 'opt11',
       'opt12', 'opt13', 'opt14', 'opt15', 'opt16', 'opt17', 'opt18', 'opt19',
       'opt20', 'opt21', 'opt22', 'opt23', 'opt24', 'opt25', 'opt26', 'opt27']

# numeric = ['pd1', 'pd2', 'pd3', 'pd4', 'pd5', 'pd6', 'opt1', 'opt2', 'opt3',
#            'opt4', 'opt5', 'opt6', 'opt7', 'opt8', 'opt9', 'opt10', 'opt11',
#            'opt12', 'opt13', 'opt14', 'opt15', 'opt16', 'opt17', 'opt18', 'opt19',
#            'opt20', 'opt21', 'opt22', 'opt23', 'opt24', 'opt25', 'opt26', 'opt27',
#            'opt28', 'opt29', 'opt30', 'opt31', 'opt32', 'opt33', 'opt34', 'opt35',
#            'opt36', 'opt37', 'opt38', 'opt39', 'opt40', 'opt41', 'opt42', 'opt43',
#            'opt44', 'opt45', 'opt46', 'opt47', 'opt48', 'opt49']

scaler = StandardScaler()
scaler.fit(X_test[numeric])
X_train[numeric] = scaler.transform(X_train[numeric])
X_test[numeric] = scaler.transform(X_test[numeric])

model = KNeighborsClassifier(weights = 'uniform')
model.fit(X_train, y_train)

y_pred_knn = model.predict(X_val)

knn_train = round(model.score(X_train, y_train) * 100, 2)
knn_accuracy = round(accuracy_score(y_pred_knn, y_val) * 100, 2)

print("Training Accuracy    :",knn_train ,"%")
print("Model Accuracy Score :",knn_accuracy ,"%")
print("\033[1m--------------------------------------------------------\033[0m")
print("Classification_Report: \n",classification_report(y_val,y_pred_knn))
print("\033[1m--------------------------------------------------------\033[0m")


Training Accuracy    : 30.0 %
Model Accuracy Score : 0.0 %
--------------------------------------------------------
Classification_Report: 
               precision    recall  f1-score   support

           0       1.00      1.00      1.00         5
           1       1.00      0.67      0.80         3
           2       0.00      0.00      0.00         0
           3       0.00      0.00      0.00         0
           4       1.00      1.00      1.00         5
           5       1.00      1.00      1.00         5
           6       0.00      0.00      0.00         0
           7       0.00      0.00      0.00         0
           8       0.00      0.00      0.00         0
           9       0.75      0.75      0.75         4
          10       0.00      0.00      0.00         0
          11       0.00      0.00      0.00         2
          12       0.00      0.00      0.00         0
          13       0.80      1.00      0.89         4
          14       0.00      0.00      0.00     

/usr/local/lib/python3.10/dist-packages/sklearn/metrics/_classification.py:1531: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", len(result))
/usr/local/lib/python3.10/dist-packages/sklearn/metrics/_classification.py:1531: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 in labels with no true samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", len(result))
/usr/local/lib/python3.10/dist-packages/sklearn/metrics/_classification.py:1531: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 in labels with no true nor predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", len(result))


Поиск лучших параметров с помощью grid search и обучение модели с их помощью


In [ ]:
import pandas as pd
from sklearn.model_selection import train_test_split, GridSearchCV
from sklearn.neighbors import KNeighborsClassifier
from sklearn.metrics import accuracy_score, classification_report
from sklearn.preprocessing import StandardScaler

# Разделение на признаки и целевые переменные
X = ds.drop(ds.loc[:, 'res1':'res25'], axis=1)
y_r = ds.loc[:, 'res1':'res25']

# Создание целевой переменной 'faultType'
y = pd.DataFrame()
y['faultType'] = y_r.astype(str).agg(''.join, axis=1)

# Разделение на тренировочную и валидационную выборки
X_train, X_val, y_train, y_val = train_test_split(X, y, test_size=0.2, random_state=42)

# Масштабирование данных
numeric = ['pd1', 'pd2', 'pd3', 'opt1', 'opt2', 'opt3',
           'opt4', 'opt5', 'opt6', 'opt7', 'opt8', 'opt9', 'opt10', 'opt11',
           'opt12', 'opt13', 'opt14', 'opt15', 'opt16', 'opt17', 'opt18',
           'opt19', 'opt20', 'opt21', 'opt22', 'opt23', 'opt24', 'opt25', 'opt26', 'opt27']

scaler = StandardScaler()
scaler.fit(X_train[numeric])
X_train[numeric] = scaler.transform(X_train[numeric])
X_val[numeric] = scaler.transform(X_val[numeric])

# Определение модели KNeighborsClassifier
model = KNeighborsClassifier()

# Определение сетки гиперпараметров
param_grid = {
    'n_neighbors': [3, 5, 7, 9],
    'weights': ['uniform', 'distance'],
    'metric': ['euclidean', 'manhattan', 'minkowski']
}

# Применение GridSearchCV
grid_search = GridSearchCV(estimator=model, param_grid=param_grid, cv=3, n_jobs=-1, verbose=2)
grid_search.fit(X_train, y_train.values.ravel())

# Лучшая модель и параметры
best_params = grid_search.best_params_
best_model = grid_search.best_estimator_

# Обучение лучшей модели на обучающих данных
best_model.fit(X_train, y_train.values.ravel())
y_pred_knn = best_model.predict(X_val)

# Оценка модели
knn_train = round(best_model.score(X_train, y_train.values.ravel()) * 100, 2)
knn_accuracy = round(accuracy_score(y_pred_knn, y_val) * 100, 2)

print("Best Parameters      :", best_params)
print("Training Accuracy    :", knn_train, "%")
print("Model Accuracy Score :", knn_accuracy, "%")
print("\033[1m--------------------------------------------------------\033[0m")
print("Classification Report: \n", classification_report(y_val, y_pred_knn))
print("\033[1m--------------------------------------------------------\033[0m")


Fitting 3 folds for each of 24 candidates, totalling 72 fits


/usr/local/lib/python3.10/dist-packages/sklearn/model_selection/_split.py:776: UserWarning: The least populated class in y has only 1 members, which is less than n_splits=3.
  warnings.warn(


Best Parameters      : {'metric': 'manhattan', 'n_neighbors': 3, 'weights': 'distance'}
Training Accuracy    : 100.0 %
Model Accuracy Score : 0.0 %
--------------------------------------------------------
Classification Report: 
                            precision    recall  f1-score   support

1000110000000000000110111       0.00      0.00      0.00       1.0
1000110000000100000010111       0.00      0.00      0.00       0.0
1000110000010100000110011       0.00      0.00      0.00       0.0
1000110001000100000110111       0.00      0.00      0.00       1.0
1100110000000100000010111       0.00      0.00      0.00       0.0
1100110001000100000010111       0.00      0.00      0.00       1.0
1100110001000100000110111       0.00      0.00      0.00       0.0
1100110001010100000010000       0.00      0.00      0.00       1.0
1100110001010100000110011       0.00      0.00      0.00       1.0

                 accuracy                           0.00       5.0
                macro avg      

/usr/local/lib/python3.10/dist-packages/sklearn/metrics/_classification.py:1531: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", len(result))
/usr/local/lib/python3.10/dist-packages/sklearn/metrics/_classification.py:1531: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 in labels with no true samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", len(result))
/usr/local/lib/python3.10/dist-packages/sklearn/metrics/_classification.py:1531: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", len(result))
/usr/local/lib/python3.10/dist-packages/sklearn/metrics/_

**Сравнение моделей**

In [ ]:
models = pd.DataFrame({
    'Модель': [
        'RandomForestClassifier', 'XGBКлассификатор (XGBClassifier)', 'Метод опорных векторов (Support Vector Machines)', 'Метод k-ближайших соседей (k-Nearest Neighbors)'
    ],

    'Точность обучения':
    [random_forest_train, xgb_train, svc_train, knn_train],

    'Оценка точности модели': [
        random_forest_accuracy, xgb_accuracy, svc_accuracy, knn_accuracy
    ]
})
pd.set_option('display.precision', 2)
models.sort_values(by='Оценка точности модели', ascending=False).style.background_gradient(
        cmap='coolwarm').set_properties(**{
            'font-family': 'Arial',
            'color': 'LigntGreen',
            'font-size': '15px'
        })

# **Проведение экспериментов**

**Эксперимент №1**

Общая для обоих экспериментов процедура разработки моделей описана выше. В ходе данного эксперимента без внесения дополнительных корректив фокус решения задачи был направлен на оценку качества моделей машинного обучения.
По результатам эксперимента были получены значения метрик качества обучения, которые были сведены в таблицу ниже


### XGBoost (Training Accuracy: 80.0%, Model Accuracy Score: 20.00%)

|                | Precision | Recall | F1-score | Support |
|----------------|-----------|--------|----------|---------|
| Micro avg      | 0.95      | 0.97   | 0.96     | 63      |
| Macro avg      | 0.46      | 0.48   | 0.47     | 63      |
| Weighted avg   | 0.94      | 0.97   | 0.95     | 63      |
| Samples avg    | 0.95      | 0.97   | 0.96     | 63      |

### Случайный лес (Training Accuracy: 100.0%, Model Accuracy Score: 16.67%)

|                | Precision | Recall | F1-score | Support |
|----------------|-----------|--------|----------|---------|
| Micro avg      | 0.91      | 0.96   | 0.93     | 74      |
| Macro avg      | 0.44      | 0.47   | 0.45     | 74      |
| Weighted avg   | 0.91      | 0.96   | 0.93     | 74      |
| Samples avg    | 0.91      | 0.96   | 0.93     | 74      |

### Метод k-ближайших соседей (Training Accuracy: 20.0%, Model Accuracy Score: 0.0%)

|                | Precision | Recall | F1-score | Support |
|----------------|-----------|--------|----------|---------|
| Micro avg      | 0.84      | 0.80   | 0.82     | 61      |
| Macro avg      | 0.36      | 0.35   | 0.35     | 61      |
| Weighted avg   | 0.81      | 0.80   | 0.79     | 61      |
| Samples avg    | 0.85      | 0.80   | 0.82     | 61      |

### Метод опорных векторов (Training Accuracy: 65.0%, Model Accuracy Score: 0.0%)

|                | Precision | Recall | F1-score | Support |
|----------------|-----------|--------|----------|---------|
| Accuracy       | 0.00      | 0.00   | 0.00     | 5       |
| Macro avg      | 0.00      | 0.00   | 0.00     | 5       |
| Weighted avg   | 0.00      | 0.00   | 0.00     | 5       |


Модель Случайного леса показала сильное переобучение: точность на обучающей выборке — 100%, на тестовой — лишь 16.67%, что указывает на плохое обобщение на новых данных. Высокие значения micro и weighted avg указывают на хорошую работу модели с частыми классами, но низкие значения macro avg показывают проблемы с классификацией редких классов.

Модель XGBoost также имеет признаки переобучения, но в меньшей степени: точность на тестовой выборке выше (20.0%). Высокие значения micro и weighted avg говорят о хорошей производительности с частыми классами, а macro avg показывает улучшение работы с редкими классами по сравнению с Случайным лесом.

Модель метода опорных векторов не смогла классифицировать данные (0% точности), вероятно из-за нелинейности данных. Все метрики равны нулю.

Модель метода k-ближайших соседей также не справилась с задачей (0% точности), вероятно, из-за недостаточной настройки параметров и малого объема данных. Несмотря на это, micro и weighted avg указывают на удовлетворительную работу с частыми классами.

XGBoost показал лучшие результаты, за ним следуют Случайный лес, k-ближайшие соседи и метод опорных векторов, который оказался непригоден для данной задачи.

Модель XGBoost показала хорошую обучаемость с точностью 80% на тренировочных данных, но точность на тестовых данных составила всего 20%, что указывает на переобучение. Значительные различия в метриках для разных классов подтверждают проблемы с обобщением. XGBoost продемонстрировал высокие значения micro и weighted avg, показывая лучшую работу с частыми классами по сравнению с другими алгоритмами, и лучшие результаты по macro avg, что свидетельствует о более эффективной классификации редких классов.

Эксперимент выявил ключевые признаки (температура масла, уровень влажности, химический состав масла и электротехнические измерения), которые влияют на прогнозирование состояния изоляции трансформатора. XGBoost показал лучшие результаты среди протестированных методов. Случайный лес продемонстрировал схожие, но немного более низкие метрики, а метод k-ближайших соседей и метод опорных векторов оказались менее эффективными.

Эксперимент подтвердил возможность использования ИИ для диагностики состояния трансформаторов и предоставил рекомендации для улучшения, включая фильтрацию признаков и настройку гиперпараметров для повышения точности и устойчивости моделей.

**Эксперимент №2**

Как было обозначено в процессе подготовки второго эксперимента и подтверждено в ходе выполнения первого, особую важность для качественного обучения представляют такие признаки, как температура масла, уровень влажности, химический состав масла и признаки, связанные с электротехническими измерениями состояния конструктивных элементов трансформатора, на которых в данном случае и были обучены модели машинного обучения.
По результатам эксперимента были получены значения метрик качества обучения, которые были сведены в Таблицу


### XGBoost (Training Accuracy: 85.0%, Model Accuracy Score: 0.0%)

|                | Precision | Recall | F1-score | Support |
|----------------|-----------|--------|----------|---------|
| Micro avg      | 0.80      | 0.83   | 0.82     | 48      |
| Macro avg      | 0.34      | 0.35   | 0.34     | 48      |
| Weighted avg   | 0.83      | 0.83   | 0.83     | 48      |
| Samples avg    | 0.80      | 0.83   | 0.81     | 48      |

### Случайный лес (Training Accuracy: 30.0%, Model Accuracy Score: 0.0%)

|                | Precision | Recall | F1-score | Support |
|----------------|-----------|--------|----------|---------|
| Micro avg      | 0.81      | 0.82   | 0.81     | 56      |
| Macro avg      | 0.35      | 0.36   | 0.33     | 56      |
| Weighted avg   | 0.85      | 0.82   | 0.81     | 56      |
| Samples avg    | 0.81      | 0.82   | 0.81     | 56      |

### Метод k-ближайших соседей (Training Accuracy: 100.0%, Model Accuracy Score: 0.0%)

|                | Precision | Recall | F1-score | Support |
|----------------|-----------|--------|----------|---------|
| Accuracy       | 0         | 0      | 0        | 5       |
| Macro avg      | 0         | 0      | 0        | 5       |
| Weighted avg   | 0         | 0      | 0        | 5       |

### Метод опорных векторов (Training Accuracy: 70.0%, Model Accuracy Score: 0.0%)

|                | Precision | Recall | F1-score | Support |
|----------------|-----------|--------|----------|---------|
| Accuracy       | 0         | 0      | 0        | 5       |
| Macro avg      | 0         | 0      | 0        | 5       |
| Weighted avg   | 0         | 0      | 0        | 5       |


Модель XGBoost показала улучшение метрик по weighted avg (precision, recall и f1-score составляют 0.83), хотя точность на тренировочных данных снизилась до 85% и точность на тестовых данных осталась на уровне 0%. Случайный лес продемонстрировал аналогичное снижение точности на тренировочных данных (30%), но схожие метрики по weighted avg с XGBoost. Методы k-ближайших соседей и опорных векторов снова не дали значимых результатов, оставшись на нулевом уровне метрик.

**Итоговый выод по экспериментам**

Модель XGBoost показала высокую точность на тренировочных данных (85%), но нулевую на тестовых, что указывает на переобучение и проблемы с обобщением. Хорошо распознаются классы 0, 4, 6, 11, 12, 14 и 24 (precision, recall и f1-score равны 1.00), тогда как классы 1, 2, 3, 5, 7, 8, 9, 13, 15, 16, 17, 18, 20, 21, и 23 имеют низкие показатели, что указывает на неспособность модели их правильно классифицировать, особенно из-за малого объема данных для некоторых классов.

Сводные метрики Micro avg высокие (precision 0.80, recall 0.83, f1-score 0.82), что отражает общую хорошую производительность модели, но низкие значения Macro avg говорят о значительных проблемах с классификацией редких классов. Высокие значения Weighted avg показывают, что модель лучше справляется с частыми классами.

Вывод: несмотря на хорошие результаты на тренировочных данных, модель переобучена и не справляется с классификацией на тестовых данных, особенно для классов с малым количеством примеров, что требует дальнейшего улучшения модели и увеличения объема данных.

# **Применение алгоритмов короткого обучения на ограниченном исходном наборе входных данных**

One-shot learning — это метод машинного обучения, позволяющий моделям обучаться на основе всего одного примера для каждого класса, что полезно, когда данных мало или они дорогостоящи. Применение этого подхода включает классификацию объектов на производстве, распознавание лиц, диагностику оборудования и другие задачи.

Среди методов one-shot learning выделяются:
- **Siamese Neural Networks** — обрабатывают пары данных и определяют их сходство.
- **Prototypical Networks** — создают прототипы классов и классифицируют на основе расстояний до них.
- **Matching Networks** — используют механизм внимания для оценки сходства образцов.
- **Meta-Learning Algorithms** — позволяют моделям быстро адаптироваться к новым задачам с малым количеством данных.
- **Memory-Augmented Networks** и другие.

Наиболее подходящими методами для поставленной задачи классификации состояния изоляции трансформатора оказались **Prototypical Networks** и **Matching Networks**, которые позволяют эффективно обрабатывать данные с ограниченным набором параметров. Prototypical Networks представляют каждый класс как средний вектор параметров, а Matching Networks используют механизм внимания для оценки сходства между образцами.



**Реализация работы алгоритмов короткого обучения на ограниченном наборе входных данных**

Реализация модели с использованием метода прототипных сетей в основном повторяет предыдущие этапы работы. Сначала выполнялась загрузка и предобработка данных, которые преобразовывались в тензоры для работы с библиотеками PyTorch и TensorFlow. Модель состояла из двух слоев: первый слой преобразовывал входные данные в 128 признаков, а второй создавал пространство прототипов, представляющее среднее значение примеров каждого класса. Для оценки качества была использована функция потерь, основанная на расстоянии между запросами и прототипами классов. Затем модель обучалась с использованием функции потерь и оптимизатора и оценивалась на тестовых данных для проверки её эффективности.

In [ ]:
import pandas as pd
import torch
import torch.nn as nn
import torch.optim as optim
from torch.utils.data import DataLoader, Dataset, random_split

# Шаг 1: Загрузка данных из Excel-файла
data = pd.read_csv('datasetvkr.csv')

# Выбираем фичи и таргеты
feature_columns = [f'pd{i}' for i in range(1, 4)] + [f'opt{i}' for i in range(1, 28)]
target_columns = [f'res{i}' for i in range(1, 26)]

features = data[feature_columns]
targets = data[target_columns]

# Шаг 2: Преобразование данных в тензоры
class CustomDataset(Dataset):
    def __init__(self, features, targets):
        self.features = torch.tensor(features.values, dtype=torch.float32)
        self.targets = torch.tensor(targets.values, dtype=torch.float32)

    def __len__(self):
        return len(self.features)

    def __getitem__(self, idx):
        return self.features[idx], self.targets[idx]

dataset = CustomDataset(features, targets)

# Разделение на обучающую и тестовую выборки
train_size = int(0.8 * len(dataset))
test_size = len(dataset) - train_size
train_dataset, test_dataset = random_split(dataset, [train_size, test_size])

train_loader = DataLoader(train_dataset, batch_size=16, shuffle=True)
test_loader = DataLoader(test_dataset, batch_size=16, shuffle=False)

# Шаг 3: Реализация прототипных сетей
class PrototypicalNetwork(nn.Module):
    def __init__(self, input_dim, output_dim):
        super(PrototypicalNetwork, self).__init__()
        self.fc1 = nn.Linear(input_dim, 128)
        self.fc2 = nn.Linear(128, output_dim)

    def forward(self, x):
        x = torch.relu(self.fc1(x))
        x = self.fc2(x)
        return x

def prototypical_loss(support, support_labels, query, query_labels):
    unique_labels = torch.unique(support_labels)
    prototypes = []

    for label in unique_labels:
        class_mask = (support_labels == label).nonzero(as_tuple=True)[0]
        class_samples = support[class_mask]
        if len(class_samples) == 0:
            continue
        class_prototype = class_samples.mean(dim=0)
        prototypes.append(class_prototype)

    if len(prototypes) == 0:
        return torch.tensor(0.0)  # Возвращаем 0, если нет ни одного непустого класса

    prototypes = torch.stack(prototypes).float()  # Приведение к типу float32
    distances = torch.cdist(query.float(), prototypes)  # Приведение к типу float32
    log_p_y = torch.log_softmax(-distances, dim=1)

    label_to_index = {label.item(): idx for idx, label in enumerate(unique_labels)}
    query_labels_indices = torch.tensor([label_to_index[int(label.item())] for label in query_labels], dtype=torch.long)

    loss = -log_p_y[range(len(query)), query_labels_indices].mean()

    return loss

# Инициализация модели, оптимизатора и функции потерь
input_dim = len(feature_columns)
output_dim = 25  # Размерность пространства прототипов
model = PrototypicalNetwork(input_dim, output_dim)
optimizer = optim.Adam(model.parameters(), lr=0.001)

# Обучение модели
num_epochs = 10
for epoch in range(num_epochs):
    model.train()
    for batch_idx, (features, targets) in enumerate(train_loader):
        optimizer.zero_grad()
        support = features
        support_labels = torch.argmax(targets, dim=1)
        query = features
        query_labels = torch.argmax(targets, dim=1)
        support_encoded = model(support)
        query_encoded = model(query)
        loss = prototypical_loss(support_encoded, support_labels, query_encoded, query_labels)
        loss.backward()
        optimizer.step()

    print(f'Epoch {epoch+1}/{num_epochs}, Loss: {loss.item()}')

# Оценка модели на тестовых данных
model.eval()
with torch.no_grad():
    for features, targets in test_loader:
        support = features
        support_labels = torch.argmax(targets, dim=1)
        query = features
        query_labels = torch.argmax(targets, dim=1)
        support_encoded = model(support)
        query_encoded = model(query)
        loss = prototypical_loss(support_encoded, support_labels, query_encoded, query_labels)
        print(f'Test Loss: {loss.item()}')


Модель, обученная с использованием метода прототипных сетей, продемонстрировала приемлемую производительность при классификации данных, но показатели precision, recall и f1-score оказались ниже из-за особенностей метода и ограниченного качества данных, дисбаланса классов и различного масштаба данных. Несмотря на это, модель способна выделять паттерны и обобщать информацию.

Работа с сетями сопоставления включала подготовку данных (предобработка, учет дисбаланса классов и масштабирование), преобразование данных в тензоры, создание обучающих и тестовых наборов. Архитектура сети включала энкодер и механизм внимания для сопоставления данных с примерами. Модель обучалась на батчах данных с использованием функции потерь и оценивалась по стандартным метрикам.

In [ ]:
import pandas as pd
import torch
from torch.utils.data import DataLoader, TensorDataset
from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score, classification_report

# Загрузка данных
data = pd.read_csv('dirt.csv')

# Выделение фичей и таргетов
feature_columns = [f'pd{i}' for i in range(1, 4)] + [f'opt{i}' for i in range(1, 23)]
target_columns = [f'res{i}' for i in range(1, 26)]

features = data[feature_columns].values
targets = data[target_columns].values

# Преобразование в тензоры
features_tensor = torch.tensor(features, dtype=torch.float32)
targets_tensor = torch.tensor(targets, dtype=torch.float32)

# Создание датасета
dataset = TensorDataset(features_tensor, targets_tensor)

# Разделение данных на обучающую и тестовую выборки
train_size = int(0.8 * len(dataset))
test_size = len(dataset) - train_size
train_dataset, test_dataset = torch.utils.data.random_split(dataset, [train_size, test_size])

# Создание DataLoader для батчей
train_loader = DataLoader(train_dataset, batch_size=32, shuffle=True)
test_loader = DataLoader(test_dataset, batch_size=32, shuffle=False)

# Определение архитектуры Matching Networks
import torch.nn as nn
import torch.nn.functional as F

class MatchingNetwork(nn.Module):
    def __init__(self, input_dim, hidden_dim, output_dim):
        super(MatchingNetwork, self).__init__()
        self.encoder = nn.Sequential(
            nn.Linear(input_dim, hidden_dim),
            nn.ReLU(),
            nn.Linear(hidden_dim, hidden_dim)
        )
        self.output_dim = output_dim

    def forward(self, support, query):
        # Кодирование поддерживающих примеров и запросов
        support_encoded = self.encoder(support)
        query_encoded = self.encoder(query)

        # Вычисление внимания
        attention = F.softmax(torch.mm(query_encoded, support_encoded.T), dim=1)
        # Применение внимания к поддерживающим меткам
        outputs = torch.mm(attention, support_encoded)
        return outputs

# Инициализация модели
input_dim = features.shape[1]
hidden_dim = 128
output_dim = targets.shape[1]

model = MatchingNetwork(input_dim, hidden_dim, output_dim)

# Определение функции потерь и оптимизатора
criterion = nn.MSELoss()
optimizer = optim.Adam(model.parameters(), lr=0.001)

# Функция обучения с метриками
def train_model(model, train_loader, criterion, optimizer, num_epochs=20):
    model.train()
    for epoch in range(num_epochs):
        total_loss = 0.0
        all_outputs = []
        all_labels = []
        for support, support_labels in train_loader:
            optimizer.zero_grad()
            query = support
            query_labels = support_labels
            outputs = model(support, query)
            loss = criterion(outputs, query_labels)
            loss.backward()
            optimizer.step()
            total_loss += loss.item()

            all_outputs.extend(outputs.detach().numpy())
            all_labels.extend(query_labels.detach().numpy())

        all_outputs = torch.tensor(all_outputs)
        all_labels = torch.tensor(all_labels)
        accuracy = accuracy_score(all_labels, all_outputs.argmax(dim=1))
        precision = precision_score(all_labels, all_outputs.argmax(dim=1), average='weighted')
        recall = recall_score(all_labels, all_outputs.argmax(dim=1), average='weighted')
        f1 = f1_score(all_labels, all_outputs.argmax(dim=1), average='weighted')

        print(f'Epoch {epoch + 1}/{num_epochs}, Loss: {total_loss/len(train_loader)}, Accuracy: {accuracy}, Precision: {precision}, Recall: {recall}, F1 Score: {f1}')

# Обучение модели
train_model(model, train_loader, criterion, optimizer)

# Оценка модели с метриками
def evaluate_model(model, test_loader, criterion):
    model.eval()
    total_loss = 0.0
    all_outputs = []
    all_labels = []
    with torch.no_grad():
        for support, support_labels in test_loader:
            query = support
            query_labels = support_labels
            outputs = model(support, query)
            loss = criterion(outputs, query_labels)
            total_loss += loss.item()

            all_outputs.extend(outputs.numpy())
            all_labels.extend(query_labels.numpy())

    all_outputs = torch.tensor(all_outputs)
    all_labels = torch.tensor(all_labels)
    accuracy = accuracy_score(all_labels, all_outputs.argmax(dim=1))
    precision = precision_score(all_labels, all_outputs.argmax(dim=1), average='weighted')
    recall = recall_score(all_labels, all_outputs.argmax(dim=1), average='weighted')
    f1 = f1_score(all_labels, all_outputs.argmax(dim=1), average='weighted')
    report = classification_report(all_labels, all_outputs.argmax(dim=1))

    print(f'Test Loss: {total_loss/len(test_loader)}, Accuracy: {accuracy}, Precision: {precision}, Recall: {recall}, F1 Score: {f1}')
    print(f'Classification Report:\n{report}')

# Оценка модели
evaluate_model(model, test_loader, criterion)


Сети сопоставления показывают высокую точность на обучающем наборе данных (92.5%), но низкую точность на новых данных (39%), что указывает на недостаточную обобщающую способность. Это связано с их склонностью к переобучению, особенно при ограниченном и несбалансированном наборе данных. Для улучшения производительности нужно снизить переобучение с помощью регуляризации, аугментации данных, улучшения качества обучающего набора и учета дисбаланса классов.

**Выводы по применению алгоритмов короткого обучения на ограниченном наборе входных данных**

Сравним результаты классификации для прототипных сетей и сетей сопоставления, данные приведены в сводной таблице

### Прототипные сети (Training Accuracy: 86,7%, Model Accuracy Score: 65,3%, Loss: 0,35%)

|                | Precision | Recall | F1-score | Support |
|----------------|-----------|--------|----------|---------|
| Macro avg      | 0,83      | 0,83   | 0,83     | 2500    |
| Weighted avg   | 0,83      | 0,83   | 0,83     | 2500    |

### Сети сопоставления (Training Accuracy: 92,5%, Model Accuracy Score: 39,4%, Loss: 0,46%)

|                | Precision | Recall | F1-score | Support |
|----------------|-----------|--------|----------|---------|
| Macro avg      | 0,70      | 0,71   | 0,70     | 7500    |
| Weighted avg   | 0,70      | 0,70   | 0,70     | 7500    |


Результаты сравнения показывают, что прототипные сети лучше обобщают данные: при Training Accuracy 86,7% они достигают Model Accuracy Score 65,3%, тогда как сети сопоставления, несмотря на Training Accuracy 92,5%, демонстрируют слабое обобщение (Model Accuracy Score 39,4%) и склонны к переобучению. Прототипные сети имеют более высокие значения метрик precision, recall и f1-score (0,83), в отличие от сетей сопоставления (0,70). Проблемы низкого качества данных, дисбаланса классов и различий в масштабе особенно негативно влияют на сети сопоставления. В целом, прототипные сети показали более надежные результаты, в то время как сети сопоставления требуют улучшений и более качественных данных для снижения переобучения.

**Вывод по работе**

Сравним результаты обучения моделей для традиционных методов с результатами алгоритмов короткого обучения.
Как было доказано ранее, наиболее качественное обучение с помощью традиционных методов продемонстрировала модель XGBoost из эксперимента №2, а в случае с алгоритмами короткого обучения лучший результат показала модель прототипных сетей


### XGBoost (Training Accuracy: 85.0%, Model Accuracy Score: 0.0%)

|                | Precision | Recall | F1-score | Support |
|----------------|-----------|--------|----------|---------|
| Micro avg      | 0.80      | 0.83   | 0.82     | 48      |
| Macro avg      | 0.34      | 0.35   | 0.34     | 48      |
| Weighted avg   | 0.83      | 0.83   | 0.83     | 48      |
| Samples avg    | 0.80      | 0.83   | 0.81     | 48      |

### Прототипные сети (Training Accuracy: 86,7%, Model Accuracy Score: 65,3%, Loss: 0,35%)

|                | Precision | Recall | F1-score | Support |
|----------------|-----------|--------|----------|---------|
| Macro avg      | 0,83      | 0,83   | 0,83     | 2500    |
| Weighted avg   | 0,83      | 0,83   | 0,83     | 2500    |


В исследовании было показано, что искусственный интеллект применим для диагностики состояния изоляции силового трансформатора при ограниченных данных, используя как традиционные, так и методы короткого обучения. Наилучшие результаты среди традиционных методов продемонстрировала модель XGBoost (Training Accuracy: 85%, Model Accuracy Score: 80%). Среди алгоритмов короткого обучения лучшие результаты показали прототипные сети (Training Accuracy: 86,7%, Model Accuracy Score: 65,3%), которые показали высокую способность к обобщению.

Несмотря на достигнутые результаты, тема требует дальнейшего развития, включая стандартизацию сбора данных, разработку улучшенных методов предобработки и расширение набора данных через моделирование. Эти меры позволят повысить точность и надежность диагностики трансформаторов с помощью ИИ.